File: _attic/bane/core/container/treapmap.go
package container

import (
	"github.com/wrmsr/bane/core/container/treap"
	rndu "github.com/wrmsr/bane/core/rand"
	bt "github.com/wrmsr/bane/core/types"
)

type treapMapComparer[K, V any] func(kv1, kv2 bt.Kv[K, V]) int

func (c treapMapComparer[K, V]) Compare(kv1, kv2 bt.Kv[K, V]) int {
	return c(kv1, kv2)
}

type TreapMap[K, V any] struct {
	n *treap.TreapNode[bt.Kv[K, V]]
	c treap.Comparer[bt.Kv[K, V]]
}

func NewTreapMap[K, V any](cmp treap.Comparer[K]) TreapMap[K, V] {
	return TreapMap[K, V]{
		c: treapMapComparer[K, V](func(v1, v2 bt.Kv[K, V]) int {
			return cmp.Compare(v1.K, v2.K)
		}),
	}
}

var _ PersistentMap[int, string] = TreapMap[int, string]{}

func (m TreapMap[K, V]) isPersistent() {}

func (m TreapMap[K, V]) Len() int {
	// TODO: memo
	result := 0
	m.n.ForEach(func(_ bt.Kv[K, V]) bool {
		result++
		return true
	})
	return result
}

func (m TreapMap[K, V]) Contains(k K) bool {
	_, ok := m.TryGet(k)
	return ok
}

func (m TreapMap[K, V]) Get(k K) V {
	if v, ok := m.TryGet(k); ok {
		return v
	}
	var z V
	return z
}

func (m TreapMap[K, V]) TryGet(k K) (V, bool) {
	n := m.n.Find(bt.KvOf[K, V](k, bt.Zero[V]()), m.c)
	if n == nil {
		var z V
		return z, false
	}
	return n.Value.V, true
}

type treapMapIterator[K, V any] struct {
	st []*treap.TreapNode[bt.Kv[K, V]]
	n  *treap.TreapNode[bt.Kv[K, V]]
	b  bool
}

var _ bt.Iterator[bt.Kv[int, string]] = &treapMapIterator[int, string]{}

func (i *treapMapIterator[K, V]) Iterate() bt.Iterator[bt.Kv[K, V]] {
	return i
}

func (i *treapMapIterator[K, V]) HasNext() bool {
	return i.n != nil
}

func (i *treapMapIterator[K, V]) Next() bt.Kv[K, V] {
	n := i.n
	if n.Right != nil {
		i.n = n.Right
		for i.n.Left != nil {
			i.st = append(i.st, i.n)
			i.n = i.n.Left
		}
	} else if len(i.st) > 0 {
		i.n = i.st[len(i.st)-1]
		i.st = i.st[:len(i.st)-1]
	} else {
		i.n = nil
	}
	return n.Value
}

func (m TreapMap[K, V]) Iterate() bt.Iterator[bt.Kv[K, V]] {
	i := &treapMapIterator[K, V]{
		st: make([]*treap.TreapNode[bt.Kv[K, V]], 0, 8),
		n:  m.n,
	}
	for i.n.Left != nil {
		i.st = append(i.st, i.n)
		i.n = i.n.Left
	}
	return i
}

func (m TreapMap[K, V]) ForEach(fn func(kv bt.Kv[K, V]) bool) bool {
	return m.n.ForEach(func(kv bt.Kv[K, V]) bool {
		return fn(kv)
	})
}

func (m TreapMap[K, V]) With(k K, v V) PersistentMap[K, V] {
	node := &treap.TreapNode[bt.Kv[K, V]]{
		bt.KvOf(k, v),
		int(rndu.FastUint32()),
		nil,
		nil,
	}
	n := m.n.Union(node, m.c, true)
	return TreapMap[K, V]{n, m.c}
}

func (m TreapMap[K, V]) Without(k K) PersistentMap[K, V] {
	n := m.n.Delete(bt.KvOf[K, V](k, bt.Zero[V]()), m.c)
	return TreapMap[K, V]{n, m.c}
}

func (m TreapMap[K, V]) Default(k K, v V) PersistentMap[K, V] {
	if _, ok := m.TryGet(k); ok {
		return m
	}
	return m.With(k, v)
}

File: omlish/collections/treapmap.py
"""
MIT License

Copyright (c) 2022 Gabriel Ochsenhofer

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
"""
import random
import typing as ta

from . import treap
from .persistent import PersistentMap


K = ta.TypeVar('K')
V = ta.TypeVar('V')


class TreapMap(PersistentMap[K, V]):
    __slots__ = ('_n', '_c')

    def __init__(
            self,
            *,
            _n: ta.Optional[treap.TreapNode[tuple[K, V]]],
            _c: treap.Comparer[tuple[K, V]],
    ) -> None:
        super().__init__()

        self._n = _n
        self._c = _c

    def __len__(self) -> int:
        # TODO: memo
        # TODO: itertools lol
        if self._n is None:
            return 0
        result = 0
        for _ in self._n:
            result += 1
        return result

    def __contains__(self, item: K) -> bool:
        try:
            self[item]  # noqa
        except KeyError:
            return False
        else:
            return True

    def __getitem__(self, item: K) -> V:
        n = treap.find(self._n, (item, None), self._c)  # type: ignore
        if n is None:
            raise KeyError(item)
        return n.value

    def __iter__(self) -> ta.Iterator[tuple[K, V]]:
        i = self.iterate()
        while i.has_next():
            yield i.next()

    def iterate(self) -> 'TreapMapIterator[K, V]':
        i = TreapMapIterator(
            _st=[],
            _n=self._n,
            _b=False,
        )
        while (n := i._n) is not None and n.left is not None:  # noqa
            i._st.append(n)  # noqa
            i._n = n.left  # noqa
        return i

    def with_(self, k: K, v: V) -> 'TreapMap[K, V]':
        node = treap.TreapNode(
            _value=(k, v),
            _priority=int(random.random() * 0xFFFFFFFF),
            _left=None,
            _right=None,
        )
        n = treap.union(self._n, node, self._c, True)
        return TreapMap(_n=n, _c=self._c)

    def without(self, k: K) -> 'TreapMap[K, V]':
        n = treap.delete(self._n, (k, None), self._c)  # type: ignore
        return TreapMap(_n=n, _c=self._c)

    def default(self, k: K, v: V) -> 'TreapMap[K, V]':
        try:
            self[k]  # noqa
        except KeyError:
            return self.with_(k, v)
        else:
            return self


def new_treap_map(cmp: ta.Callable[[tuple[K, V], tuple[K, V]], int]) -> PersistentMap[K, V]:
    return TreapMap(_n=None, _c=cmp)


class TreapMapIterator(ta.Generic[K, V]):
    __slots__ = ('_st', '_n', '_b')

    def __init__(
            self,
            *,
            _st: list[treap.TreapNode[tuple[K, V]]],
            _n: ta.Optional[treap.TreapNode[tuple[K, V]]],
            _b: bool,
    ) -> None:
        super().__init__()

        self._st = _st
        self._n = _n
        self._b = _b

    def has_next(self) -> bool:
        return self._n is not None

    def next(self) -> tuple[K, V]:
        n = self._n
        if n is None:
            raise StopIteration
        if n.right is not None:
            self._n = n.right
            while self._n.left is not None:
                self._st.append(self._n)
                self._n = self._n.left
        elif len(self._st) > 0:
            self._n = self._st[-1]
            self._st.pop()
        else:
            self._n = None
        return n.value
